src/tapedfunction.jl

```diff
@@ -56,72 +25,132 @@ mutable struct TapedFunction{F}
         end
 
         ir = CodeInfoTools.code_inferred(f, args_type...)
-        tf = new{F}() # leave some fields to be undef
-        tf.func, tf.arity, tf.ir = f, length(args), ir
-        tf.tape = RawTape()
-        tf.counter = 1
+        tape = RawTape()
+        utape = Vector{FunctionWrapper{Nothing, Tuple{TapedFunction}}}()
+        bindings = translate!(tape, ir)
 
-        translate!(tf, ir)
+        tf = new{F}(f, length(args), ir, tape, utape, 1, bindings, :none)
         TRCache[cache_key] = tf # set cache
+        # unify!(tf)
         return tf
     end
 
     function TapedFunction(tf::TapedFunction{F}) where {F}
-        new{F}(tf.func, tf.arity, tf.ir, tf.tape,
+        new{F}(tf.func, tf.arity, tf.ir, tf.tape, tf.unified_tape,
                tf.counter, tf.bindings, :none)
     end
 end
 
 const TRCache = LRU{Tuple, TapedFunction}(maxsize=10)
 
+function unify!(tf::TapedFunction)
+    length(tf.tape) == length(tf.unified_tape) && return
+    for ins in tf.tape
+        push!(tf.unified_tape, FunctionWrapper{Nothing, Tuple{TapedFunction}}(ins))
+    end
+end
+
+# const TypedFunction = FunctionWrapper
+struct TypedFunction{OT, IT<:Tuple}
+    func::Function
+    retval::Ref{OT}
```

Comment:
This should be changed to
```suggestion
    retval::Base.RefValue{OT}
```
since
```julia
julia> isconcretetype(Ref{Float64})
false

julia> isconcretetype(Base.RefValue{Float64})
true
```
This makes a huge difference in practice, e.g.
```julia
julia> using BenchmarkTools

julia> struct A{T}
           x::Ref{T}
       end

julia> struct B{T}
           x::Base.RefValue{T}
       end

julia> function f(x::Union{A{T},B{T}}) where {T}
           s = zero(T)
           for _ in 1:10_000
               s += x.x[]
           end
           s
       end
f (generic function with 1 method)

julia> @btime f($(A(Ref(1.0))));
  130.273 μs (20000 allocations: 312.50 KiB)

julia> @btime f($(B(Ref(1.0))));
  8.492 μs (0 allocations: 0 bytes)
```